deb/openmediavault-bcache/usr/share/php/openmediavault/system/storage/bcache/backingdevice.inc

```diff
@@ -0,0 +1,257 @@
+	/**
+	 * Creates a backing device.
+	 * @return void
+	 * @throw \OMV\ExecException
+	 */
+	public function create() {
+		$cmdArgs = [];
+		$cmdArgs[] = "-B";
+		$cmdArgs[] = $this->getDeviceFile();
+		$cmd = new \OMV\System\Process("make-bcache", $cmdArgs);
+		$cmd->setRedirect2to1();
+		$cmd->execute();
+		while (!$this->isRegistered()) {
+			sleep(1);
+		}
+	}
+
+	/**
+	 * Enable caching with the given cache device.
+	 * @return void
+	 * @throw \OMV\ExecException
+	 */
+	public function attach($csetUuid) {
+		$this->setValue("attach", $csetUuid);
+	}
+
+	/**
+	 * Get the device file of the bcache device.
+	 * @return string Returns the device file.
+	 * @throw \OMV\ExecException
+	 */
+	public function getBcacheDeviceFile() {
+		$path = "/sys/block/{$this->getDeviceName(TRUE)}/bcache/dev";
+		if (!file_exists($path)) {
+			return "";
+		}
+		$file = new \SplFileInfo($path);
+		// link target: ../../../../../../../../../../../virtual/block/bcache0
+		$linkTarget = $file->getLinkTarget();
+		$arr = explode(DIRECTORY_SEPARATOR, $linkTarget);
+		return sprintf("/dev/%s", $arr[count($arr) - 1]);
+	}
+
+	/**
+	 * Get the device file of the cache device.
+	 * @return string Returns the device file.
+	 * @throw \OMV\ExecException
+	 */
+	public function getCacheDeviceFile() {
+		$path = sprintf("/sys/block/%s/bcache/cache/cache0",
+			$this->getDeviceName(TRUE));
+		if (!file_exists($path)) {
+			return "";
+		}
+		$file = new \SplFileInfo($path);
+		// link target: ../../../devices/virtual/block/<device name>/bcache
+		$linkTarget = $file->getLinkTarget();
+		$arr = explode(DIRECTORY_SEPARATOR, $linkTarget);
+		return sprintf("/dev/%s", $arr[count($arr) - 2]);
+	}
+
+	/**
+	 * The backing device can be in one of four different states:
+	 * no cache: Has never been attached to a cache set.
+	 * clean: Part of a cache set, and there is no cached dirty data.
+	 * dirty: Part of a cache set, and there is cached dirty data.
+	 * inconsistent: The backing device was forcibly run by the user
+	 *   when there was dirty data cached but the cache set was
+	 *   unavailable; whatever data was on the backing device has
+	 *   likely been corrupted.
+	 * @return string The state.
+	 * @throw \OMV\ExecException
+	 */
+	public function getState() {
+		if (!$this->isRegistered()) {
+			return 'unregistered';
+		}
+		if (!$this->isRunning()) {
+			return 'missing cache';
+		}
+		return $this->getValue("state");
+	}
+
+	/**
+	 * If bcache is running (i.e. whether the /dev/bcache device exists,
+	 * whether it’s in passthrough mode or caching.
+	 * @return boolean True if bcache is running, otherwise false.
+	 * @throw \OMV\ExecException
+	 */
+	public function isRunning() {
+		return $this->getValue("running") == 1;
+	}
+
+	/**
+	 * Shut down the bcache device and close the backing device.
+	 * @return void
+	 * @throw \OMV\ExecException
+	 */
+	public function stop() {
+		while ($this->isRegistered()) {
+			$this->setValue("stop", 1);
```

Comment:
Is it really necessary to call this method on every iteration?

Reply:
I think once should be enough.